File: TypeScript Solutions/295.ts
class MedianFinder {
  private arr: number[] = [];
  constructor() {}

  // binary search, O(logn)
  addNum(num: number): void {
    const binarySearch = (n: number) => {
      let start = 0;
      let end = this.arr.length;
      while (start < end) {
        let mid = Math.floor((start + end) / 2);
        if (n > this.arr[mid]) start = mid + 1;
        else end = mid;
      }
      // insert the number at start (mid)
      this.arr.splice(start, 0, n);
    };
    if (this.arr.length === 0) this.arr.push(num);
    else binarySearch(num);
  }

  // O(1)
  findMedian(): number {
    if (this.arr.length % 2 === 0) {
      const mid = Math.floor(this.arr.length / 2) - 1;
      return (this.arr[mid] + this.arr[mid + 1]) / 2;
    } else {
      const mid = Math.floor(this.arr.length / 2);
      return this.arr[mid];
    }
  }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * var obj = new MedianFinder()
 * obj.addNum(num)
 * var param_2 = obj.findMedian()
 */

File: Java Solutions/295.java

// https://leetcode.com/problems/find-median-from-data-stream/discuss/74047/JavaPython-two-heap-solution-O(log-n)-add-O(1)-find
import java.util.Collections;
import java.util.PriorityQueue;

class MedianFinder {
  // largest at heap head
  private PriorityQueue<Integer> small = new PriorityQueue<>(Collections.reverseOrder());
  // smallest at heap head (natural order)
  private PriorityQueue<Integer> large = new PriorityQueue<>();
  private boolean even = true;

  public MedianFinder() {

  }

  public void addNum(int num) {
    if (even) {
      // to keep the 2 pq even size
      large.offer(num);
      small.offer(large.poll());
    } else {
      small.offer(num);
      large.offer(small.poll());
    }
    even = !even;
  }

  public double findMedian() {
    if (even) {
      return (small.peek() + large.peek()) / 2.0;
    } else {
      return small.peek();
    }
  }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */